vlm/src/main/scala/geotrellis/contrib/vlm/package.scala

```diff
@@ -69,6 +75,118 @@ package object vlm {
     new StreamingByteReader(rr)
   }
 
+  implicit class tokenMethods(val token: Long) extends AnyVal {
+
+    def getProjection(): Option[String] = {
+      val crs = Array.ofDim[Byte](1 << 16)
+      GDALWarp.get_crs_wkt(token, 0, crs)
+      Some(new String(crs, "UTF-8"))
+    }
+
+    def rasterExtent(): RasterExtent = {
+      val transform = Array.ofDim[Double](6)
+      val width_height = Array.ofDim[Int](2)
+      GDALWarp.get_transform(token, 0, transform)
+      GDALWarp.get_width_height(token, 0, width_height)
+
+      val x1 = transform(0)
+      val y1 = transform(3)
+      val x2 = x1 + transform(1) * width_height(0)
+      val y2 = y1 + transform(5) * width_height(1)
+      val e = Extent(
+        math.min(x1, x2),
+        math.min(y1, y2),
+        math.max(x1, x2),
+        math.max(y1, y2))
+
+      RasterExtent(e,
+        math.abs(transform(1)), math.abs(transform(5)),
+        width_height(0), width_height(1))
+    }
+
+    def resolutions: List[RasterExtent] = {
+      val N = 1 << 8;
+      val widths = Array.ofDim[Int](N)
+      val heights = Array.ofDim[Int](N)
+      val extent = token.extent
+      GDALWarp.get_overview_widths_heights(token, 0, widths, heights)
+      widths.zip(heights).flatMap({ case (w, h) =>
+        if (w > 0 && h > 0) Some(RasterExtent(extent, cols = w, rows = h))
+        else None
+      }).toList
+    }
+
+    def extent(): Extent = token.rasterExtent.extent
+
+    def bandCount(): Int = {
+      val count = Array.ofDim[Int](1)
+      GDALWarp.get_band_count(token, 0, count)
+      count(0)
+    }
+
+    def crs(): CRS = {
+      val crs = Array.ofDim[Byte](1 << 16)
+      GDALWarp.get_crs_proj4(token, 0, crs)
+      val proj4String: String = new String(crs, "UTF-8").trim
+      if (proj4String.length > 0) CRS.fromString(proj4String.trim)
+      else LatLng
+    }
+
+    def noDataValue(): Option[Double] = {
+      val nodata = Array.ofDim[Double](1)
+      val success = Array.ofDim[Int](1)
+      GDALWarp.get_band_nodata(token, 0, 1, nodata, success)
+      if (success(0) == 0)
+        None
+      else
+        Some(nodata(0))
+    }
+
+    def dataType(): Int = {
+      val dataType = Array.ofDim[Int](1)
+      GDALWarp.get_band_data_type(token, 0, 1, dataType)
+      dataType(0)
+    }
+
+    def cellSize(): CellSize = {
+      val transform = Array.ofDim[Double](6)
+      GDALWarp.get_transform(token, 0, transform)
+      CellSize(transform(1), transform(5))
+    }
+
+    def cellType(): CellType = {
```

Comment:
In reality GeoTrellis and the TIFF lib have a bit richer types for TIFFs. 
GDAL erases some information (for instance a sign of values, [tiff tag 339](https://www.awaresystems.be/imaging/tiff/tifftags/sampleformat.html))

That's how we convert GDAL type into the GeoTrellis type (not perfect since GDAL doesn't allow us to get access to the TIFF TAG ): https://github.com/geotrellis/geotrellis-gdal/blob/b81508039d26999e8a829a2ff432e9132cb24169/gdal/src/main/scala/geotrellis/gdal/GDALUtils.scala#L39

[That's how we do it in GeoTrellis Core reader](https://github.com/locationtech/geotrellis/blob/ff27a4654f3e488c3313a8ab52069fe3f16f8f78/raster/src/main/scala/geotrellis/raster/io/geotiff/reader/GeoTiffReader.scala#L258) 

In reality in the link above `BandType` is defined by the `bits per sample tag` and `sample format`. GDAL provides a `pits per sample` thing but not a `sample format`. [Tags actually GeoTrellis reader supports](https://github.com/locationtech/geotrellis/blob/ff27a4654f3e488c3313a8ab52069fe3f16f8f78/raster/src/main/scala/geotrellis/raster/io/geotiff/tags/codes/TagCodes.scala#L77).

It can be pretty important to keep behaviour of the original reader and the new one as similar as it's possible.

Reply:
Okay, that is helpful, thanks.

Reply:
Okay, using GDALUtils, now.